craft_providers/executor.py

```diff
@@ -0,0 +1,101 @@
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Executor module."""
+import logging
+import pathlib
+import subprocess
+from abc import ABC, abstractmethod
+from typing import List
+
+logger = logging.getLogger(__name__)
+
+
+class Executor(ABC):
+    """Interfaces to execute commands and move data in/out of an environment."""
+
+    @abstractmethod
+    def create_file(
+        self,
+        *,
+        destination: pathlib.Path,
+        content: bytes,
+        file_mode: str,
+        group: str = "root",
+        user: str = "root",
+    ) -> None:
+        """Create file with content and file mode.
+
+        :param destination: Path to file.
+        :param content: Contents of file.
+        :param file_mode: File mode string (e.g. '0644').
+        :param group: File owner group ID.
+        :param user: Filer owner user ID.
+        """
+
+    @abstractmethod
+    def execute_popen(self, command: List[str], **kwargs) -> subprocess.Popen:
+        """Execute command in instance, using subprocess.Popen().
+
+        :param command: Command to execute.
+        :param kwargs: Additional keyword arguments to pass.
+
+        :returns: Popen instance.
+        """
+
+    @abstractmethod
+    def execute_run(
+        self, command: List[str], check=True, **kwargs
+    ) -> subprocess.CompletedProcess:
+        """Execute command using subprocess.run().
+
+        :param command: Command to execute.
+        :param kwargs: Keyword args to pass to subprocess.run().
+        :param check: Raise exception on failure.
+
+        :returns: Completed process.
+
+        :raises subprocess.CalledProcessError: if command fails and check is
+            True.
+        """
+
+    @abstractmethod
+    def pull_file(self, *, source: pathlib.Path, destination: pathlib.Path) -> None:
+        """Copy environment file to host destination.
+
+        Providing this as an abstract method allows the provider to implement
+        the most performant option available.
+
+        :param source: Environment file to copy.
+        :param destination: Host path to copy to.
+
+        :raises FileNotFound: If source file or destination directory does not
+            exist.
+        :raises OSError: On error copying file.
+        """
+
+    @abstractmethod
+    def push_file(self, *, source: pathlib.Path, destination: pathlib.Path) -> None:
+        """Copy host file into environment at destination.
```

Comment:
Same than above for destionation being a dir.

(for both) all this implies I can not do `.push_file(source=Path('/tmp/foo'), destination=Path('/tmp/foo')`?

Reply:
Yes thanks.